qiskit/circuit/quantumcircuit.py

```diff
@@ -1939,7 +1939,13 @@ def _assign_parameter(self, parameter, value):
                 replace instances of ``parameter``.
         """
         for instr, param_index in self._parameter_table[parameter]:
-            instr.params[param_index] = instr.params[param_index].assign(parameter, value)
+            new_param = instr.params[param_index].assign(parameter, value)
+            # if fully bound, validate
+            if len(new_param.parameters) == 0:
+                instr.params[param_index] = instr.validate_parameter(new_param)
+            else:
+                instr.params[param_index] = new_param
```

Comment:
Is this redundant from the `validate_parameters` call in the setter for `Instruction.params`?

Reply:
The setter is only called if one does `instr.params = params` (i.e. sets all parameters at once). If you do item assignment, the getter is called and then `__setitem__` on the list. It would be nicer if the validation would also be called then, but that would require some additional logic we don't have at the moment.